sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py

```diff
@@ -21,6 +22,85 @@
+
+        # Get the scheme of the challenge and remove it from the challenge string
+        trimmed_challenge = self._get_challenge_parts(challenge)
+        split_challenge = trimmed_challenge.split(" ", 1)
+        self.scheme = split_challenge[0]
+        trimmed_challenge = split_challenge[1]
+
+        # Split trimmed challenge into name=value pairs; these pairs are expected to be split by either commas or spaces
+        # Values may be surrounded by quotes, which are stripped here
+        separator = "," if "," in trimmed_challenge else " "
+        for item in trimmed_challenge.split(separator):
+            # process name=value pairs
+            comps = item.split("=")
+            if len(comps) == 2:
+                key = comps[0].strip(' "')
+                value = comps[1].strip(' "')
+                if key:
+                    self._parameters[key] = value
+
+        if not self._parameters:
+            raise ValueError("Invalid challenge parameters")
+
+        # Challenge must specify authorization or authorization_uri
+        if "authorization" not in self._parameters and "authorization_uri" not in self._parameters:
+            raise ValueError("Invalid challenge parameters")
+
+        authorization_uri = self.get_authorization_server()
+        # the authorization server URI should look something like https://login.windows.net/tenant-id[/oauth2/authorize]
+        uri_path = urllib.parse.urlparse(authorization_uri).path.lstrip("/")
+        self.tenant_id = uri_path.split("/")[0] or None
+
+    def get_authorization_server(self):
+        """ Returns the URI for the authorization server if present, otherwise an empty string. """
+        value = ""
+        for key in ["authorization_uri", "authorization"]:
+            value = self._parameters.get(key) or ""
+            if value:
+                break
+        return value
+
+    def get_resource(self):
+        """ Returns the resource if present, otherwise an empty string. """
+        return self._parameters.get("resource") or self._parameters.get("resource_id") or ""
+
+    def get_scope(self):
+        """ Returns the scope if present, otherwise an empty string. """
+        return self._parameters.get("scope") or ""
+
+    # pylint:disable=no-self-use
+    def _get_challenge_parts(self, challenge):
+        """ Verifies that the challenge is a valid auth challenge and returns the key=value pairs. """
+        if not challenge:
+            raise ValueError("Challenge cannot be empty")
+
+        return challenge.strip()
+
+    # pylint:disable=no-self-use
+    def _get_request_authority(self, uri):
+        """ Extracts the host authority from the given URI. """
+        if not uri:
+            raise ValueError("request_uri cannot be empty")
+
+        uri = urllib.parse.urlparse(uri)
+        if not uri.netloc:
+            raise ValueError("request_uri must be an absolute URI")
+
+        if uri.scheme.lower() not in ["http", "https"]:
```

Comment:
Realistically would it ever not be one of these? And if it isn't... something probably would have failed sooner?
Not really requesting any changes, but I'm curious if this validation was actually driven by a particular scenario?

Reply:
This was written [three years ago](https://github.com/Azure/azure-sdk-for-python/blame/be9bc2ebd94b204d4939b848ec9b23688eeb12c3/sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/_shared/http_challenge.py#L107) for Key Vault, so I doubt even Charles would remember at this point... I can't personally think of a real scenario where this would come up, though.